File: flutter/lib/features/weather_search/data/mappers/weather_forecast_mapper.dart
import '../../domain/entities/weather_exception.dart';
import '../../domain/entities/weather_forecast.dart';
import '../models/weather_forecast_response.dart';

/// 將 API 的 `LocationData`（各要素平行時段陣列）轉為以「時段」為主的領域實體。
class WeatherForecastMapper {
  const WeatherForecastMapper();

  /// 缺少必要要素或各要素時段數不一致 / 為 0 時，拋出
  /// `WeatherException(WeatherErrorType.invalidData)`。
  WeatherForecast map(LocationData location) {
    final elementsByName = <String, List<TimeData>>{
      for (final element in location.weatherElements) element.elementName: element.times,
    };

    final wx = elementsByName['Wx'];
    if (wx == null || wx.isEmpty) {
      throw const WeatherException(WeatherErrorType.invalidData);
    }
    final periodCount = wx.length;

    // 取出與 Wx 時段數對齊的要素，缺少或不對齊即視為資料格式錯誤。
    List<TimeData> aligned(String name) {
      final times = elementsByName[name];
      if (times == null || times.length != periodCount) {
        throw const WeatherException(WeatherErrorType.invalidData);
      }
      return times;
    }

    final pop = aligned('PoP');
    final minT = aligned('MinT');
    final maxT = aligned('MaxT');
    final ci = aligned('CI');

    final periods = List<WeatherPeriod>.generate(periodCount, (index) {
      return WeatherPeriod(
        startTime: wx[index].startTime,
        endTime: wx[index].endTime,
        weatherDescription: wx[index].parameterName,
        rainProbability: pop[index].parameterName,
        minTemperature: minT[index].parameterName,
        maxTemperature: maxT[index].parameterName,
        comfort: ci[index].parameterName,
      );
    });

    return WeatherForecast(cityName: location.locationName, periods: periods);
  }
}

File: ios-native/Packages/Sources/DataModule/WeatherSearch/Mapper/WeatherForecastMapper.swift
import DomainModule
import Foundation

/// 將 API 的 `Location`（各要素平行時段陣列）轉為以「時段」為主的領域模型。
enum WeatherForecastMapper {
    /// - Throws: ``WeatherError/invalidData`` 當缺少必要要素或各要素時段數不一致 / 為 0。
    static func map(_ location: WeatherForecastResponse.Location) throws -> WeatherForecastModel {
        var elementsByName: [String: [WeatherForecastResponse.TimePeriod]] = [:]
        for element in location.weatherElement {
            elementsByName[element.elementName] = element.time
        }

        guard let wx = elementsByName["Wx"], !wx.isEmpty else {
            throw WeatherError.invalidData
        }
        let periodCount = wx.count

        // 取出與 Wx 時段數對齊的要素，缺少或不對齊即視為資料格式錯誤。
        func aligned(_ name: String) throws -> [WeatherForecastResponse.TimePeriod] {
            guard let times = elementsByName[name], times.count == periodCount else {
                throw WeatherError.invalidData
            }
            return times
        }

        let pop = try aligned("PoP")
        let minT = try aligned("MinT")
        let maxT = try aligned("MaxT")
        let ci = try aligned("CI")

        let periods = (0..<periodCount).map { index in
            WeatherPeriodModel(
                startTime: wx[index].startTime,
                endTime: wx[index].endTime,
                weatherDescription: wx[index].parameter.parameterName,
                rainProbability: pop[index].parameter.parameterName,
                minTemperature: minT[index].parameter.parameterName,
                maxTemperature: maxT[index].parameter.parameterName,
                comfort: ci[index].parameter.parameterName
            )
        }
        return WeatherForecastModel(cityName: location.locationName, periods: periods)
    }
}
